Ricl web3

lzl124631x opened the pull request:
- [x] hmy_getBlockByNumber
- [x] hmy_getBlockByHash
- [x] hmy_getBlockTransactionCountByHash
- [x] hmy_getBlockTransactionCountByNumber
- [x] hmy_getTransactionByBlockHashAndIndex
- [x] hmy_getTransactionByBlockNumberAndIndex
- [x] hmy_getTransactionByHash
- [x] hmy_getCode
- [x] hmy_getStorageAt
- [x] hmy_getBalance
- [x] hmy_getTransactionCount

The others I'll punt them for next PR.
- [ ] hmy_sendTransaction
- [ ] hmy_sendRawTransaction
- [ ] hmy_syncing
- [ ] net_peerCount

chaosma commented on rpc/hmyapi/public.go:
```diff
@@ -0,0 +1,236 @@
+	b *rpc.HmyAPIBackend
+}
+
+// NewPublicBlockChainAPI creates a new Ethereum blockchain API.
```
create a new Harmony blockchain API

chaosma commented on rpc/hmyapi/public.go:
```diff
@@ -0,0 +1,236 @@
+func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash common.Hash, fullTx bool) (map[string]interface{}, error) {
+	block, err := s.b.GetBlock(ctx, blockHash)
+	if block != nil {
+		return RPCMarshalBlock(block, false, false)
```
Does this method need to nil out some fields for pending block? or pending block will has no blockhash returned

lzl124631x replied:
No, according to https://github.com/ethereum/go-ethereum/blob/ee376f6766ab32616dbdeca0872691305002230b/internal/ethapi/api.go#L587

LeoHChen commented on rpc/http.go:
```diff
@@ -0,0 +1,339 @@
+//
+// Deprecated: Server implements http.Handler
+func NewHTTPServer(srv http.Handler) *http.Server {
+	// Wrap the CORS-handler within a host-handler
```
remove debug code

lzl124631x commented on api/service/rpc/service.go:
```diff
@@ -0,0 +1,126 @@
+	apis := hmyapi.GetAPIs(rpc.NewBackend(s.blockchain, s.txPool))
+
+	port, _ := strconv.Atoi(s.peer.Port)
+	s.httpEndpoint = fmt.Sprintf("127.0.0.1:%v", port+rpcPortDiff)
```
Making the default host localhost is good enough (eth doing the same https://github.com/ethereum/go-ethereum/blob/92a849a5095a3d53b8c7d52be32c583ef2acdc6d/node/defaults.go#L31). Later if we want different host I'll add a parameter.

lzl124631x commented:
> This PR is too huge. We should really distinguish between our own code and code copied from other places. Otherwise, it is very hard to maintain a stable code base going forward.

They are quite necessary, including json marshalling, ipc and websocket supports. Files in `hmyapi` are what I mainly edited. Others I applied changed on demand.

LeoHChen commented:
it is important to keep track of the original source of the files for RPC.  So that we can keep updating and keep an eye on the bug fixes.  Can you do two things?  First of all, just copy all the original files from external source and keep a record of the version.  Second, apply your changes on top of the original ones.  So that we know what's the original copies, and what's your particular changes.  If you have done these two things, I can approve the merge of the PR.

lzl124631x commented:
If I separate them, the first set of changes cannot build.
Also I fixed lots of TSlint issues of eth code I don't want to redo it again.
Even if I split them, they will be squashed eventually.
If you take a look at my commit history, you will see the first commit "Port" are mainly ported code. The subsequent changes are for implementing them in the right way.

Files that are mainly ported over (might include tslint fix):
* rpc/client.go: the rpc client.
* rpc/endpoints.go: starting HTTP/WS/IPC endpoints.
* rpc/errors.go: error code handling.
* rpc/handler.go: handling requests, running methods.
* rpc/http.go: starting HTTP service.
* rpc/ipc_unix.go: util for ipc unix 
* rpc/ipc.go: util for general ipc
* rpc/json.go: json de-/serialization
* rpc/server.go: the rpc server
* rpc/stdio.go: stdio for ipc.
* rpc/subscription.go: ws subscriptions.
* rpc/types.go: type declarations
* rpc/websocket.go: ws connection.
* rpc/hmyapi/addrlock.go

My changes:
* rpc/service.go: our service (http/ws/ipc) starter/stoper
* other files under `rpc/hmyapi/`.
